File: Trees/SizeBalancedTreeMethods2.cs
using System;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Platform.Collections.Methods.Trees
{
    /// <summary>
    /// Experimental implementation, don't use it yet.
    /// </summary>
    public unsafe abstract class SizeBalancedTreeMethods2<TElement> : SizedBinaryTreeMethodsBase<TElement>
    {
        protected override void AttachCore(IntPtr root, TElement newNode)
        {
            if (ValueEqualToZero(root))
            {
                System.Runtime.CompilerServices.Unsafe.Write((void*)root, newNode);
                IncrementSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root));
            }
            else
            {
                IncrementSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root));
                if (FirstIsToTheLeftOfSecond(newNode, System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)))
                {
                    AttachCore(GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)), newNode);
                    LeftMaintain(root);
                }
                else
                {
                    AttachCore(GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)), newNode);
                    RightMaintain(root);
                }
            }
        }

        protected override void DetachCore(IntPtr root, TElement nodeToDetach)
        {
            if (ValueEqualToZero(root))
            {
                return;
            }
            var currentNode = root;
            var parent = IntPtr.Zero; /* Изначально зануление, так как родителя может и не быть (Корень дерева). */
            var replacementNode = GetZero();
            while (!IsEquals(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode), nodeToDetach))
            {
                SetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode), Decrement(GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode))));
                if (FirstIsToTheLeftOfSecond(nodeToDetach, System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode)))
                {
                    parent = currentNode;
                    currentNode = GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode));
                }
                else if (FirstIsToTheRightOfSecond(nodeToDetach, System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode)))
                {
                    parent = currentNode;
                    currentNode = GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)currentNode));
                }
                else
                {
                    throw new InvalidOperationException("Duplicate link found in the tree.");
                }
            }
            if (!ValueEqualToZero(GetLeftPointer(nodeToDetach)) && !ValueEqualToZero(GetRightPointer(nodeToDetach)))
            {
                var minNode = GetRightValue(nodeToDetach);
                while (!EqualToZero(GetLeftValue(minNode)))
                {
                    minNode = GetLeftValue(minNode); /* Передвигаемся до минимума */
                }
                DetachCore(GetRightPointer(nodeToDetach), minNode);
                SetLeft(minNode, GetLeftValue(nodeToDetach));
                if (!ValueEqualToZero(GetRightPointer(nodeToDetach)))
                {
                    SetRight(minNode, GetRightValue(nodeToDetach));
                    SetSize(minNode, Increment(Add(GetSize(GetLeftValue(nodeToDetach)), GetSize(GetRightValue(nodeToDetach)))));
                }
                else
                {
                    SetSize(minNode, Increment(GetSize(GetLeftValue(nodeToDetach))));
                }
                replacementNode = minNode;
            }
            else if (!ValueEqualToZero(GetLeftPointer(nodeToDetach)))
            {
                replacementNode = GetLeftValue(nodeToDetach);
            }
            else if (!ValueEqualToZero(GetRightPointer(nodeToDetach)))
            {
                replacementNode = GetRightValue(nodeToDetach);
            }
            if (parent == IntPtr.Zero)
            {
                System.Runtime.CompilerServices.Unsafe.Write((void*)root, replacementNode);
            }
            else if (IsEquals(GetLeftValue(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)parent)), nodeToDetach))
            {
                SetLeft(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)parent), replacementNode);
            }
            else if (IsEquals(GetRightValue(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)parent)), nodeToDetach))
            {
                SetRight(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)parent), replacementNode);
            }
            ClearNode(nodeToDetach);
        }

        private void LeftMaintain(IntPtr root)
        {
            if (!ValueEqualToZero(root))
            {
                var rootLeftNode = GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root));
                if (!ValueEqualToZero(rootLeftNode))
                {
                    var rootRightNode = GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root));
                    var rootLeftNodeLeftNode = GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootLeftNode));
                    if (!ValueEqualToZero(rootLeftNodeLeftNode) &&
                        (ValueEqualToZero(rootRightNode) || GreaterThan(GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootLeftNodeLeftNode)), GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootRightNode)))))
                    {
                        RightRotate(root);
                    }
                    else
                    {
                        var rootLeftNodeRightNode = GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootLeftNode));
                        if (!ValueEqualToZero(rootLeftNodeRightNode) &&
                            (ValueEqualToZero(rootRightNode) || GreaterThan(GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootLeftNodeRightNode)), GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootRightNode)))))
                        {
                            LeftRotate(GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)));
                            RightRotate(root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    LeftMaintain(GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)));
                    RightMaintain(GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)));
                    LeftMaintain(root);
                    RightMaintain(root);
                }
            }
        }

        private void RightMaintain(IntPtr root)
        {
            if (!ValueEqualToZero(root))
            {
                var rootRightNode = GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root));
                if (!ValueEqualToZero(rootRightNode))
                {
                    var rootLeftNode = GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root));
                    var rootRightNodeRightNode = GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootRightNode));
                    if (!ValueEqualToZero(rootRightNodeRightNode) &&
                        (ValueEqualToZero(rootLeftNode) || GreaterThan(GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootRightNodeRightNode)), GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootLeftNode)))))
                    {
                        LeftRotate(root);
                    }
                    else
                    {
                        var rootRightNodeLeftNode = GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootRightNode));
                        if (!ValueEqualToZero(rootRightNodeLeftNode) &&
                            (ValueEqualToZero(rootLeftNode) || GreaterThan(GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootRightNodeLeftNode)), GetSize(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)rootLeftNode)))))
                        {
                            RightRotate(GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)));
                            LeftRotate(root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    LeftMaintain(GetLeftPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)));
                    RightMaintain(GetRightPointer(System.Runtime.CompilerServices.Unsafe.Read<TElement>((void*)root)));
                    LeftMaintain(root);
                    RightMaintain(root);
                }
            }
        }
    }
}

File: Platform.Collections.Methods.Console/SizeBalancedTreeMethods2.cpp


namespace Platform::Collections::Methods::Trees
{
    template <typename TElement> class SizeBalancedTreeMethods2 : public SizedBinaryTreeMethodsBase<TElement>
    { public:

        virtual TElement* GetLeftReference(TElement node) override = 0;

        virtual TElement* GetRightReference(TElement node) override = 0;

        virtual TElement GetLeft(TElement node) override = 0;

        virtual TElement GetRight(TElement node) override = 0;

        virtual TElement GetSize(TElement node) override = 0;

        virtual void SetLeft(TElement node, TElement left) override = 0;

        virtual void SetRight(TElement node, TElement right) override = 0;

        virtual void SetSize(TElement node, TElement size) override = 0;

        virtual bool FirstIsToTheLeftOfSecond(TElement first, TElement second) override = 0;

        virtual bool FirstIsToTheRightOfSecond(TElement first, TElement second) override = 0;

        void AttachCore(TElement* root, TElement node) override
        {
            if (*root == 0)
            {
                *root = node;
                SizedBinaryTreeMethodsBase<TElement>::IncrementSize(*root);
            }
            else
            {
                SizedBinaryTreeMethodsBase<TElement>::IncrementSize(*root);
                if (FirstIsToTheLeftOfSecond(node, *root))
                {
                    AttachCore(GetLeftReference(*root), node);
                    LeftMaintain(root);
                }
                else
                {
                    AttachCore(GetRightReference(*root), node);
                    RightMaintain(root);
                }
            }
        }

        void DetachCore(TElement* root, TElement nodeToDetach) override
        {
            TElement* currentNode = root;
            TElement* parent = root;
            TElement replacementNode = 0;
            while (currentNode != nodeToDetach)
            {
                SetSize(currentNode, GetSize(currentNode) - 1);
                if (FirstIsToTheLeftOfSecond(nodeToDetach, currentNode))
                {
                    parent = currentNode;
                    currentNode = GetLeftReference(currentNode);
                }
                else if (FirstIsToTheRightOfSecond(nodeToDetach, currentNode))
                {
                    parent = currentNode;
                    currentNode = GetRightReference(currentNode);
                }
                else
                {
                    throw std::exception("Duplicate link found in the tree.");
                }
            }
            TElement nodeToDetachLeft = GetLeft(nodeToDetach);
            TElement node = GetRight(nodeToDetach);
            if (nodeToDetachLeft != 0 && node != 0)
            {
                TElement minNode = node;
                TElement minNodeLeft = GetLeft(minNode);
                while (minNodeLeft != 0)
                {
                    minNode = minNodeLeft;
                    minNodeLeft = GetLeft(minNode);
                }
                DetachCore(GetRightReference(nodeToDetach), minNode);
                SetLeft(minNode, nodeToDetachLeft);
                node = GetRight(nodeToDetach);
                if (node != 0)
                {
                    SetRight(minNode, node);
                    SetSize(minNode, GetSize(nodeToDetachLeft) + GetSize(node) + 1);
                }
                else
                {
                    SetSize(minNode, GetSize(nodeToDetachLeft) + 1);
                }
                replacementNode = minNode;
            }
            else if (nodeToDetachLeft != 0)
            {
                replacementNode = nodeToDetachLeft;
            }
            else if (node != 0)
            {
                replacementNode = node;
            }
            if (*root == nodeToDetach)
            {
                *root = replacementNode;
            }
            else if (GetLeft(parent) == nodeToDetach)
            {
                SetLeft(parent, replacementNode);
            }
            else if (GetRight(parent) == nodeToDetach)
            {
                SetRight(parent, replacementNode);
            }
            SizedBinaryTreeMethodsBase<TElement>::ClearNode(nodeToDetach);
        }

        void LeftMaintain(TElement* root)
        {
            if (*root != 0)
            {
                TElement rootLeftNode = GetLeft(*root);
                if (rootLeftNode != 0)
                {
                    TElement rootRightNode = GetRight(*root);
                    TElement rootLeftNodeLeftNode = GetLeft(rootLeftNode);
                    if (rootLeftNodeLeftNode != 0 &&
                        (rootRightNode == 0 || GetSize(rootLeftNodeLeftNode) > GetSize(rootRightNode)))
                    {
                        SizedBinaryTreeMethodsBase<TElement>::RightRotate(root);
                    }
                    else
                    {
                        TElement rootLeftNodeRightNode = GetRight(rootLeftNode);
                        if (rootLeftNodeRightNode != 0 &&
                            (rootRightNode == 0 || GetSize(rootLeftNodeRightNode) > GetSize(rootRightNode)))
                        {
                            SizedBinaryTreeMethodsBase<TElement>::LeftRotate(GetLeftReference(*root));
                            SizedBinaryTreeMethodsBase<TElement>::RightRotate(root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    LeftMaintain(GetLeftReference(*root));
                    RightMaintain(GetRightReference(*root));
                    LeftMaintain(root);
                    RightMaintain(root);
                }
            }
        }

        void RightMaintain(TElement* root)
        {
            if (*root != 0)
            {
                TElement rootRightNode = GetRight(*root);
                if (rootRightNode != 0)
                {
                    TElement rootLeftNode = GetLeft(*root);
                    TElement rootRightNodeRightNode = GetRight(rootRightNode);
                    if (rootRightNodeRightNode != 0 &&
                        (rootLeftNode == 0 || GetSize(rootRightNodeRightNode) > GetSize(rootLeftNode)))
                    {
                        SizedBinaryTreeMethodsBase<TElement>::LeftRotate(root);
                    }
                    else
                    {
                        TElement rootRightNodeLeftNode = GetLeft(rootRightNode);
                        if (rootRightNodeLeftNode != 0 &&
                            (rootLeftNode == 0 || GetSize(rootRightNodeLeftNode) > GetSize(rootLeftNode)))
                        {
                            SizedBinaryTreeMethodsBase<TElement>::RightRotate(GetRightReference(*root));
                            SizedBinaryTreeMethodsBase<TElement>::LeftRotate(root);
                        }
                        else
                        {
                            return;
                        }
                    }
                    LeftMaintain(GetLeftReference(*root));
                    RightMaintain(GetRightReference(*root));
                    LeftMaintain(root);
                    RightMaintain(root);
                }
            }
        }
    };
}